subprojects/parseq-guava-interop/src/main/java/com/linkedin/parseq/guava/ListenableFutureUtil.java

```diff
@@ -41,19 +41,16 @@ public static <T> Task<T> fromListenableFuture(ListenableFuture<T> future) {
         new BaseTask<T>("fromListenableFuture: " + Task._taskDescriptor.getDescription(future.getClass().getName())) {
           @Override
           public boolean cancel(Exception rootReason) {
-            if (future.isCancelled()) {
-              boolean taskCancelResult = super.cancel(rootReason);
-              LOGGER.warn("Future is cancelled, therefore cancelling task, result: {}", taskCancelResult);
-              return taskCancelResult;
+            // <BaseTask>.cancel()'s result indicates whether cancel() successfully trigger state transition to "CANCELLED"
+            // And we should only cancel GRPC future when the transition was conducted.
+            boolean baseStateTransitedByCallingCancel = super.cancel(rootReason);
+            if (baseStateTransitedByCallingCancel && !future.isCancelled()) {
+              boolean futureCancelResult = future.cancel(true);
+              if (!futureCancelResult) {
+                LOGGER.error("Unexpected error: GRPC future was not cancelled but new attempt to cancel also failed.");
```

Comment:
I don't think that we should error for this. We tried to cancel gRPC future, but if it failed, we should not error here, at most log as warning.

Reply:
+1 for warning. Cancellation is best effort, and if the future is already complete, it is likely that it cannot be cancelled